### PR TITLE
Completed Orders Routes and controllers

### LINKED ISSUE #15: orders controller and routes
Can you give me information about the json data that is expected while calling the API when creating a new order.

### PR DESCRIPTION
**Pull request by ganesh-dagadi**

### CODE REVIEW

File: controllers/order.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+})
+
+// @route GET /orders/:id
+// @descp Get an Order
+// @access Private
+exports.getAnOrder = asyncHandler(async (req , res , next)=>{
+    const orderId = req.params.id;
+    const foundOrder = await Order.findById(orderId);
+    if(!foundOrder){
+        res.status(500).json({error : "something went wrong"});
```
Comment (fetahokey) [inline controllers/order.js]: I think it's 404 error in here
`res.status(404).json({error : "resource could not be found !"});`

File: controllers/order.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        res.status(200).json({order : foundOrder})
+    }
+})
+
+// @route PUT /orders/:id
+// @descp Edit an order
+// @access Private
+exports.updateOrder = asyncHandler(async (req , res , next)=>{
+    const orderId = req.params.id;
+    updatedOrder = await Order.findByIdAndUpdate(orderId , req.body);
```
Comment (fetahokey) [inline controllers/order.js:41]: you can check if the `order `exists else return a `404 `error before proceeding next step

File: controllers/order.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        res.status(200).json({order : updatedOrder})
+    }
+})
+
+// @route DELETE /orders/:id
+// @descp Delete an Order
+// @access Private
+exports.deleteOrder = asyncHandler(async (req , res , next)=>{
+    const orderId = req.params.id;
+    console.log(orderId)
```
Comment (fetahokey) [inline controllers/order.js]: delete unused console.log

File: controllers/order.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    const deletedOrder = await Order.findByIdAndDelete(orderId);
+    console.log(deletedOrder)
```
Comment (fetahokey) [inline controllers/order.js]: delete unused console.log

File: controllers/order.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    }
+})
+
+// @route DELETE /orders/:id
+// @descp Delete an Order
+// @access Private
+exports.deleteOrder = asyncHandler(async (req , res , next)=>{
+    const orderId = req.params.id;
+    console.log(orderId)
+    const deletedOrder = await Order.findByIdAndDelete(orderId);
```
Comment (fetahokey) [inline controllers/order.js:58]: I think checking for the resource existence before applying for any database action is more convenient

File: controllers/order.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+})
+
+// @route GET /orders/find/:userId
+// @descp Create a new Order
+// @access Private
+exports.getUserOrders = asyncHandler(async(req , res , next)=>{
+    const userId = req.params.userId;
+    const orders = await Order.find({user : userId})
+    if(!orders){
+        res.status(500).json({error : "Something went wrong"})
```
Comment (fetahokey) [inline controllers/order.js:66]: you can get rid of this, it will return an empty array `[]`

File: controllers/order.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    }
+})
+
+// @route GET /orders/
+// @descp Get all orders
+// @access Private (Admin)
+exports.getAllOrders = asyncHandler(async(req , res , next)=>{
+    const orders = await Order.find()
+    if(!orders){
+        res.status(500).json({error : "Something went wrong"})
```
Comment (fetahokey) [inline controllers/order.js]: you can get rid of this, it will return an empty array []

File: controllers/order.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    try{
+        const income = await Order.aggregate([
+            {$match : {createdAt : {$gte : previousMonth}}},
+            {$project : {month : {$month : "$createdAt"} , sales : "$amount"}},
+            {$group : {_id : "$month", total : {$sum : "$sales"}}}
+        ])
+        res.status(200).json(income)
+    }catch(err){
+        res.status(500).json(err)
+    }
```
Comment (fetahokey) [inline controllers/order.js]: when we use `asyncHandler`, almost you don't need to `try-catch`  block

File: routes/auth.js [hunk 9]
```diff
@@ -20,7 +20,7 @@ router.post("/", validateRegisterUser, registerUser);
 // load user
 router.get("/:id/user", verifyTokenAndAdmin, getUser);
 
-// get all useres
+// get all users
```
Comment (fetahokey) [inline routes/auth.js:23]: 👍

File: utils/verifyToken.js [hunk 10]
```diff
@@ -21,7 +21,7 @@ const verifyTokenAndAuthorization = (req, res, next) => {
     if (req.user.id === req.params.id || req.user.isAdmin) {
       return next();
     } else {
-      return res.status(403).json({ error: "Your not allowd for that!" });
+      return res.status(403).json({ error: "Your not allowed to do that!" });
```
Comment (fetahokey) [inline utils/verifyToken.js:24]: 👍

File: validators/orders.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    check("products" , "Products cant be empty").not().isEmpty(),
+    check("amount", "Amount is missing").not().isEmpty(),
+    check("address", "Address is missing").not().isEmpty(),
+    (req, res, next) => {
+        const errors = validationResult(req);
+        if (!errors.isEmpty())
+          return res.status(400).json({ error: errors.array() });
+        next();
+    },
+]
```
Comment (fetahokey) [inline validators/orders.js:14]: good start 🎉

**Review by fetahokey — CHANGES_REQUESTED** [review]
Good start, the Orders routes look like a very interesting addition 🚀 💯 Left a few suggestions here and there.

 - The overall formatting of your PR, helps a lot with reviews and project management! One tip would be to set a self-explanatory title using plain English (e.g. Implement routes for orders ' .... ). You can also check out this [article ](https://hugooodias.medium.com/the-anatomy-of-a-perfect-pull-request-567382bb6067) if you're interested as it goes into more detail about best practices when formatting PRs.

File: controllers/order.js (hunk 1 above)
Comment (fetahokey) [reply controllers/order.js]: HTTP Status Codes > [ref ](https://restfulapi.net/http-status-codes/#:~:text=404%20(Not%20Found),condition%20is%20temporary%20or%20permanent.)

File: controllers/order.js (hunk 2 above)
Comment (ganesh-dagadi) [reply controllers/order.js:41]: Do you mean if OrderId is falsy, I return a `404` error?

Comment (fetahokey) [reply controllers/order.js:41]: yes I think the case, you receive a false/invalid `orderId ` to update `Order`, doesn't mean internal server error

**Pull request closed without merging**